src/main/java/net/dv8tion/jda/api/components/utils/ComponentPathIterator.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2015 Austin Keener, Michael Ritter, Florian Spieß, and the JDA contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.dv8tion.jda.api.components.utils;
+
+import net.dv8tion.jda.api.components.Component;
+import net.dv8tion.jda.api.components.actionrow.ActionRow;
+import net.dv8tion.jda.api.components.container.Container;
+import net.dv8tion.jda.api.components.section.Section;
+import org.apache.commons.collections4.iterators.SingletonIterator;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * Utility class to recursively iterate on a list of components while keeping track their path.
+ */
+public class ComponentPathIterator implements Iterator<ComponentPathIterator.ComponentWithPath>
+{
+    private final Stack<Iterator<ComponentWithPath>> stack = new Stack<>();
```

Comment:
Use `ArrayDeque` instead of the obsolete stack, we don't need synchronization on an iterator.